### DISCUSSION TITLE
Memory leaks with BullMQ sandboxed workers

### DISCUSSION THREAD

**Discussion by tolgap** [mikro-orm/mikro-orm#2511] [Q&A]
I have a custom application that uses BullMQ workers in sandboxed mode. I have 2 queues in use. These 2 queues are handled asynchronously in separate processes.

I'm having trouble figuring out how to handle the DB connection. To be precise, it's more about when to close the DB. Right now, I'm using the following logic in my BullMQ processor entry point:

```ts
export default async function Processor(job: Job<BaseJobData>) {
  // create inversify container with all repo's and service classes that inject each other.
  const container = await initContainer();
  const connection = container.get<MikroORM<IDatabaseDriver<Connection>>>(
    TYPES.DATABASE_CONNECTION
  );

  // create the RequestContext so we don't use the global EM instance. Within this context, we execute the BullMQ Job.
  await RequestContext.createAsync(connection.em, async function () {
    await performJob(job);
  });

  // I would like to call `connection.close(true)` here so the worker process can exit gracefully
  await connection.close(true);
}
```

But when I call `connection.close()` at the end of my processor, it seems that BullMQ can "reuse" this process for another Job in the Queue that is waiting. And then, I have a DB connection which has been closed, and cannot be reused again.

What logic would be necessary to not have dangling worker processes that keep a connection alive to the DB in the context of sandboxed BullMQ workers? Anyone got tips?

**Comment by B4nan** [comment]
I guess this is a wrong place to ask questions like that, seems like bullmq specific thing (as opposed to MikroORM specific - you could ask the same thing with anything, e.g. raw redis connection, no ORM needed and you will have the same problem). From MikroORM point of view, you have a method to close the connection, and that's it, what more would you expect?

To me this sounds like you should not initialize the DI container inside processor if it can be reused, there should be better place to init and teardown dependencies then.

**Reply by tolgap** [reply]
Thank you for taking the time to respond. You're right, it seems more like a BullMQ issue. For anyone running into this, I have created a discussion at BullMQ: https://github.com/taskforcesh/bullmq/discussions/942
